src/audio/mixin_mixout.c

```diff
@@ -248,34 +243,31 @@ static void mixout_free(struct comp_dev *dev)
 	rfree(dev);
 }
 
-static
-struct mixout_source_info *find_mixout_source_info(struct mixed_data_info __sparse_cache *mdi,
-						   const struct comp_dev *mixin)
-{
-	/* mixin == NULL is also a valid input -- this will find first unused entry */
-	int i;
-
-	for (i = 0; i < MIXOUT_MAX_SOURCES; i++) {
-		if (mdi->source_info[i].mixin == mixin)
-			return &mdi->source_info[i];
-	}
-
-	return NULL;
-}
-
-/* Clears counters for not yet produced frames in mixout sink buffer. Preserves binded mixins
- * pointers as these pointers should only be cleared on unbind event.
+/* Clears counters for not yet produced frames in mixout sink buffer.
  * This function is intended to be called on mixout sink buffer stream reset, i.e.,
  * in mixout_prepare().
  */
-static void reset_mixed_data_info_frame_counters(struct mixed_data_info __sparse_cache *mdi)
+static void reset_mixed_data_info_frame_counters(struct comp_dev *mixout_dev)
 {
-	int i;
+	struct mixed_data_info __sparse_cache *mixed_data_info;
+	struct mixout_data *md = comp_get_drvdata(mixout_dev);
+	struct list_item *blist;
 
-	mdi->mixed_frames = 0;
+	mixed_data_info = mixed_data_info_acquire(md->mixed_data_info);
+	mixed_data_info->mixed_frames = 0;
+	mixed_data_info_release(mixed_data_info);
+
+	list_for_item(blist, &mixout_dev->bsource_list) {
+		struct comp_buffer *unused_in_between_buf;
+		struct mixin_data *mixin_data;
+		struct comp_dev *mixin;
 
-	for (i = 0; i < MIXOUT_MAX_SOURCES; i++)
-		mdi->source_info[i].consumed_yet_not_produced_frames = 0;
+		unused_in_between_buf = buffer_from_list(blist, struct comp_buffer,
+							 PPL_DIR_UPSTREAM);
+		mixin = buffer_get_comp(unused_in_between_buf, PPL_DIR_UPSTREAM);
+		mixin_data = comp_get_drvdata(mixin);
+		mixin_data->consumed_yet_not_produced_frames = 0;
```

Comment:
maybe we should put a comment here to say that also this is protected by IPC serialisation